### ISSUE TITLE
model expected the shape of dimension 0 to be between 1 and 1 but received 32

### ISSUE THREAD

**Issue by wanduoz** [triton-inference-server/server#4445]
I want to use dynamic batching in triton plan model. I use pytorch hub resnet50 pretrained model and convert to onnx. 
![image](https://user-images.githubusercontent.com/33130310/170957960-a0a1ec7c-40fc-48e1-aad2-08491275e968.png)

Model input and output information
![image](https://user-images.githubusercontent.com/33130310/170958432-2a3ae61e-17ee-4c2e-a64c-4ae6753c42bb.png)

Then I use trtexec to convert onnx model into tensorrt. 
```
trtexec --onnx=resnet50_onnx_bs.onnx \
--saveEngine=model.plan \
--explicitBatch \
--minShapes=input:1x3x224x224 \
--optShapes=input:8x3x224x224 \
--maxShapes=input:32x3x224x224
```

In order to deploy model at triton server, I follow the instructions and directory is 
![image](https://user-images.githubusercontent.com/33130310/170959411-058d92b8-05d3-43b8-bb86-2f33524d629c.png)

Config file
![image](https://user-images.githubusercontent.com/33130310/170959871-a4d06bd4-d198-4119-8dfa-794675e20988.png)

Start triton server
![image](https://user-images.githubusercontent.com/33130310/170959981-36a2991e-7cf6-4555-9e31-b1ed62b60ed1.png)

Client code
```
import time
import numpy as np
import tritonclient.http as httpclient
from PIL import Image

if __name__ == "__main__":
    triton_client = httpclient.InferenceServerClient(url=127.0.0.1:8000')

    image = Image.open('./dog.jpg')
    image = image.resize((224, 224), Image.ANTIALIAS)
    image = np.asarray(image)
    image = image / 255
    image = np.expand_dims(image, axis=0)
    image = np.array(np.repeat(image, 32, axis=0), dtype=np.float32)
    image = np.transpose(image, axes=[0, 3, 1, 2])
    image = image.astype(np.float32)
    print("input shape: ", image.shape)  # 32,3,224,224
    inputs = []
    inputs.append(httpclient.InferInput('input', image.shape, "FP32"))
    inputs[0].set_data_from_numpy(image, binary_data=False)
    outputs = []
    outputs.append(httpclient.InferRequestedOutput('output', binary_data=False)) 
    results = triton_client.infer('trt-resnet50', inputs=inputs, outputs=outputs)
```

Unfortunately, I get error information.
```
root@sib-1080Ti_x2-225:/models# python client.py
client.py:12: DeprecationWarning: ANTIALIAS is deprecated and will be removed in Pillow 10 (2023-07-01). Use Resampling.LANCZOS instead.
  image = image.resize((224, 224), Image.ANTIALIAS)
input shape:  (32, 3, 224, 224)
Traceback (most recent call last):
  File "client.py", line 58, in <module>
    results = triton_client.infer('trt-resnet50', inputs=inputs, outputs=outputs)
  File "/usr/local/lib/python3.8/dist-packages/tritonclient/http/__init__.py", line 1372, in infer
    _raise_if_error(response)
  File "/usr/local/lib/python3.8/dist-packages/tritonclient/http/__init__.py", line 64, in _raise_if_error
    raise error
tritonclient.utils.InferenceServerException: request specifies invalid shape for input 'input' for trt-resnet50. Error details: model expected the shape of dimension 0 to be between 1 and 1 but received 32
```

I use similar configuration for pytorch/onnx/openvino, but they can output result successfully. My questions are 

1. Does tensorrt support dynamic batch and how to solve above error.
2. Can I set "execution_accelerators" in onnx config as an alternative. Something like this.
```
optimization { execution_accelerators {
  gpu_execution_accelerator : [ {
    name : "tensorrt"
    parameters { key: "precision_mode" value: "FP16" }
    parameters { key: "max_workspace_size_bytes" value: "1073741824" }
    }]
}}
```

**Comment by Tabrizian** [comment]
The issue is coming from the model itself. I think in your scenario you don't need to use explicit batch and should build the model with implicit batch. (cc @tanmayv25 to correct me if I'm wrong here).

**Comment by wanduoz** [comment]
> The issue is coming from the model itself. I think in your scenario you don't need to use explicit batch and should build the model with implicit batch. (cc @tanmayv25 to correct me if I'm wrong here).

Thanks for replying. I try the belowing command
```
trtexec --onnx=resnet50_onnx_bs.onnx \
--saveEngine=model.plan \
--minShapes=input:1x3x224x224 \
--optShapes=input:8x3x224x224 \
--maxShapes=input:32x3x224x224
```

Then I use the same client code, but still get dimension error
```
tritonclient.utils.InferenceServerException: request specifies invalid shape for input 'input' for trt-resnet50. Error details: model expected the shape of dimension 0 to be between 1 and 1 but received 32
```

(By the way, I don't know how cc someone. Do you mean @ someone?)

---

There are still questions bothering me.

1. When I use onnx or torchScript model in triton, Can I just set **optimization tensorrt** and the triton will conver model into tensorrt plan file(the 2nd question I asked above)? If so what is the difference between manually convert model and automatically convert model?
2. Back to topic of dynamic batching, can I just batch data at client side and send it to server or I need to simulate concurrent scenario(start 32 processes and each send one image data)? Can triton handle both?
3. I want to estimate the inference performance and I calculate time interval before and after infer. But It cose about 4s. Do you have any idea about this?
```
start = time.time()
results = triton_client.infer('trt-resnet50', inputs=inputs, outputs=outputs)
end = time.time()
print("triton trt infer: ", (end-start), "s")
```

**Comment by GuanLuo** [comment]
Can you examine your TRT model as described [here](https://www.ccoderun.ca/programming/doxygen/tensorrt/md_TensorRT_tools_Polygraphy_examples_cli_inspect_02_inspecting_a_tensorrt_engine_README.html) and see if it has the correct profiles (for dynamic shape)?

> When I use onnx or torchScript model in triton, Can I just set optimization tensorrt and the triton will conver model into tensorrt plan file(the 2nd question I asked above)? If so what is the difference between manually convert model and automatically convert model?

Some backends does support online conversion (i.e. ORT backend has TRT execution accelarator, also see [here](https://github.com/triton-inference-server/server/blob/main/docs/optimization.md#framework-specific-optimization)). I don't have clear pointer to the difference, but in general, online conversion usually results in longer model load time and can be sub-optimal.

> Back to topic of dynamic batching, can I just batch data at client side and send it to server or I need to simulate concurrent scenario(start 32 processes and each send one image data)? Can triton handle both?

Yes, each request can have different batch size (< max batch size) and Triton will try to form largest possible batch in dynamic batching.

> I want to estimate the inference performance and I calculate time interval before and after infer. But It cose about 4s. Do you have any idea about this?

Note that the communication between client and server has to go through network, which can introduce overhead when sending large tensor and / or the network bandwidth is small. You may use [perf analyzer](https://github.com/triton-inference-server/server/blob/main/docs/perf_analyzer.md) which provides a more detailed breakdown on time spent in different stage of the inference.

**Comment by tanmayv25** [comment]
@wanduoz  As Iman pointed, the issue is coming from the model. The generated model does not appear to be dynamic.

Can you share the following output for the model.plan that you generated using trtexec?
```
polygraphy inspect model model.plan
```

> 1. When I use onnx or torchScript model in triton, Can I just set optimization tensorrt and the triton will conver model into tensorrt plan file(the 2nd question I asked above)? If so what is the difference between manually convert model and automatically convert model?

You can use tensorrt optimization in onnx and torchscript model. But the models will still run via ORT and PyTorch backend respectively. The TRT optimization will acccelerate the inference but it all comes from the ORT and TorchScript frameworks. For even better performance we recommend you use TRT models via TRT backend. Which is more scalable and performant.
For instance, for every new shape ORT will trigger a new engine generation.

>  2. Back to topic of dynamic batching, can I just batch data at client side and send it to server or I need to simulate concurrent scenario(start 32 processes and each send one image data)? Can triton handle both?

You can use both, The second one can be obrtained using [dynamic batching](https://github.com/triton-inference-server/server/blob/main/docs/model_configuration.md#dynamic-batcher).

> 3. I want to estimate the inference performance and I calculate time interval before and after infer. But It cose about 4s. Do you have any idea about this?

I would advice using perf_analyzer for better perf analysis.

**Comment by wanduoz** [comment]
@tanmayv25 @GuanLuo Thank you for your replies. I follow your instructions and make several experiments. ( I use another model so the shape changes).

1. use `polygraphy inspect model model.plan` to check the model structure. I guess -1 represent dynamic batch here and this is what I want. Then I use use triton to test. 
![image](https://user-images.githubusercontent.com/33130310/171833672-99b0edea-e340-4189-b89c-3f8c5092757e.png)

2. the configuration are as follows
![image](https://user-images.githubusercontent.com/33130310/171834234-a1d54441-3b03-4da3-a8c8-649b92fbd28f.png)

3. I send batchsize=4 random data, and it return the result successfully.
![image](https://user-images.githubusercontent.com/33130310/171834477-6d6ca4b4-10ee-443c-bc77-8091100f3c07.png)

4. However, when I change bs=16 in the client code. It raise an error.
`tritonclient.utils.InferenceServerException: request specifies invalid shape for input 'input0' for trt-pt-dynamicBs_0. Error details: model expected the shape of dimension 0 to be between 1 and 4 but received 16`

5. Here is my opinion. Triton support dynamic batching. Although my model supports a maximum batch size of 4, triton will automatically breakdown the whole data into many chunks of data whose bs equals 4. But the reality is that user need to set a max bs when convert onnx to trt and the maximum bs in config.pbtxt and data shouldn't greater than aforementioned value. Can I modify model or config to satisfy my request?

---

I also make so expriment about model performance using `perf_analyzer`. It turns out that trt has much better performance than torchScript and onnx. But the result of openvino model seems abnormal. So I just examine the time cost in the client code. How this happen?

(client code)
![image](https://user-images.githubusercontent.com/33130310/171837830-c74170b1-f1c4-4388-a099-ce11511e18b4.png)

`
result
jit-pt infer time:  1055.3584098815918  ms
onnx-pt infer time:  2128.530263900757  ms
openvino-pt infer time:  137.88747787475586  ms
trt-pt infer time:  31.87394142150879  ms
`

(torchScript performance)
![image](https://user-images.githubusercontent.com/33130310/171837181-d43cc4ae-42dd-476d-a823-83c3c05b2a99.png)

(trt performance)
![image](https://user-images.githubusercontent.com/33130310/171837040-3fa7d189-dec1-40f8-babc-20ffff2b6d0b.png)

(openvino performance)
![image](https://user-images.githubusercontent.com/33130310/171836825-ab3a8d5b-083d-4dfe-959a-87092bb722b4.png)

**Comment by tanmayv25** [comment]
> Here is my viewpoint. Triton support dynamic batching. Although my model supports a maximum batch size of 4, triton will automatically breakdown the whole data into many chunks of data whose bs equals 4. But it turns out that I need to set a max bs when convert onnx to trt and the maximum bs in config.pbtxt and data shouldn't greater than aforementioned value. Can I modify model or config to satisfy my request? (client send bs=16 data to server and server split it into 4 pieces)

This is not how it works. Please read [here ](https://github.com/triton-inference-server/server/blob/main/docs/model_configuration.md#dynamic-batcher)for more information about Triton's dynamic batching. Triton only grows the batch and not breaks it into smaller batches. The client must make sure the batch size of their request is less than or equal to the max_batch_size supported by the model. Otherwise server returns an error as you observed.

OpenVINO models run on CPU-only and does not utilize GPU. Hence it makes sense that the performance might be lesser than TRT. 

However, I don't understand the below values:

```
jit-pt infer time:  1055.3584098815918  ms 
onnx-pt infer time:  2128.530263900757  ms 
```

Can you double check your measurements here?

**Comment by wanduoz** [comment]
@tanmayv25 You can ignore this two results. I have 4 models, torchScript, onnx, openvino and tensorrt. I just test the time interval before and after calling `triton_client.infer()`. Openvino and tensorrt cost less time than torhcScript which is reasonable. But the result of `perf_analyzer` seems abnormal and that is what i confuse about.

**Comment by tanmayv25** [comment]
@wanduoz  Can you summarize what numbers I should be looking at? And how are the perf_analyzer numbers abnormal. 

If I am looking at perf_analyzer data that you shared so far

| Performance      | Perf Analyzer | Python Client(single request) |
| ----------- | ----------- | ----------- |
| torchScript      | 18ms       | **1055 ms ???**       |
| openVino   | 131 ms        |  137ms        |
| TensorRT   | 13ms        |  31 ms        |

To me the numbers for TorchScript looks abnormal.

**Comment by tanmayv25** [comment]
This is more as what I would expect. Running inference -- in this case a convolution operation -- is much faster on GPU than on a CPU. 
Comparing openvino backend's performance that currently only supports slower CPU executions to others that run on GPU with higher compute parallelism should give similar results.

**Comment by tanmayv25** [comment]
Closing as the original issue is resolved.
